pageserver/src/walingest.rs

```diff
@@ -305,6 +305,17 @@ impl<'a> WalIngest<'a> {
                     self.checkpoint_modified = true;
                 }
             }
+        } else if decoded.xl_rmid == pg_constants::RM_LOGICALMSG_ID {
+            let info = decoded.xl_info & pg_constants::XLR_RMGR_INFO_MASK;
+            if info == pg_constants::XLOG_LOGICAL_MESSAGE {
+                // This is a convenient way to make the WAL ingestion pause at
+                // particular point in the WAL. For more fine-grained control,
+                // we could peek into the message and only pause if it contains
+                // a particular string, for example, but this is enough for now.
+                info!("processing logical message");
+                fail::fail_point!("wal-ingest-logical-message");
+                info!("processing logical message: done");
+            }
         }
```

Comment:
In the test, you are using the `sleep(...)` failpoint action.
That uses `std::thread::sleep()` internally.
But this is an async function.
So, it blocks the executor thread.

Please use `utils::failpoint_sleep_millis_async` and `return(milliseconds_to_sleep)` as a failpoint action in the test.


```suggestion
            let info = decoded.xl_info & pg_constants::XLR_RMGR_INFO_MASK;
            if info == pg_constants::XLOG_LOGICAL_MESSAGE {
                // This is a convenient way to make the WAL ingestion pause at
                // particular point in the WAL. For more fine-grained control,
                // we could peek into the message and only pause if it contains
                // a particular string, for example, but this is enough for now.
                info!("processing logical message");
                utils::failpoint_sleep_millis_async!(("wal-ingest-logical-message-sleep");
                info!("processing logical message: done");
            }
        }
```